### ISSUE TITLE
Create histogram of agent wealth at end of simulation

### ISSUE THREAD

**Issue by WheezePuppet** [WheezePuppet/specnet#17]
Create histogram of agent wealth at end of simulation.

**Comment by venkatachalapathy** [comment]
As we discussed today, we need to justify what it means to *end* the simulation.

**Comment by divilian** [comment]
Yeah, good pt. Let's bring the others (@IngenuityArts, @srikanthmudigonda, @jeffg828) in to this.

My paradigm in ABM has typically been "run the simulation for a fixed number of iterations, then end it and take stock of its state at that point." As I understand it, this is what the SriMil model did. Rajesh brought up a good point in a Zoom session yesterday, though: this seems pretty arbitrary. More principled than "run for 1000 iterations," perhaps, would be "run until some sort of 'equilibrium' is encountered," however that is measured.

I agree with Rajesh in principle. The Devil's in the details: how does one define "steady state" or "equilibrium?" Certainly if it means "the state of the simulation at time _n_ and time _n_+1 are identical," that's never going to happen if (say) random real-valued quantities are added to the 'wealth' of each agent every time step. So one could back off from that and say "if the number of proto institutions, and the membership in each, is unchanged over some number of time periods, we consider that 'equilibrium.'" The "some number of time periods" part is still arbitrary, though, and in any case, this misses what could be an important dynamic: if (say) some kinds of agents tend for form protos earlier, and hence benefit from having done so, that seems important to capture. Just waiting until everyone eventually forms one doesn't tell you the important thing (unless long-term wealth were impacted by earlier formation, which it probably will be, but still).

Thoughts?

**Comment by jeffg828** [comment]
Does the model generate any visual output?

**Comment by venkatachalapathy** [comment]
It would be nice to have an animation where one can see histogram's evolution over time. I know a few python and R packages that have this capacity.  For now, here is a link to [`matplotlib` histogram animation](https://matplotlib.org/2.1.2/gallery/animation/histogram.html). We could call Python via `PyCall` julia library.

**Comment by divilian** [comment]
> Does the model generate any visual output?

Currently, SPECnet generates a dynamic animation of the graph, like [this one](http://cs.umw.edu/~stephen/output.gif). (Recall: the relative wealth of an agent is proportional to its node size; lines between nodes represent social connections between agents; pairs of colored nodes indicate proto-institutions; black nodes represent "dead" agents that have fallen below zero wealth and therefore no longer interact.)

I'm happy to add any other visualization stuff we think is useful for analysis; we just need to figure out what.

**Comment by jeffg828** [comment]
I was asking because a visualization is an easy way to see if an
equilibrium has been reached.  An equilibrium isn't necessarily when
nothing is changing any more, but could be that the creation and
destruction of stuff has balanced out.  A graph makes it easy to see this.

**Comment by divilian** [comment]
> I was asking because a visualization is an easy way to see if an equilibrium has been reached. An equilibrium isn't necessarily when nothing is changing any more, but could be that the creation and destruction of stuff has balanced out. A graph makes it easy to see this.

Ah, I see. Yeah, good point -- now we need to figure out exactly what the graph would show. For instance, if it showed "the number of protos over time," then presumably the line would asymptote to some number, at which point we could do the Jeff thing and say "ah yes, it looks like at iteration 1027 or so there were no longer any changes to the number of protos." I guess we have to figure out what our equilibrium criteria are exactly, so we know what kind of graph to create.

**Comment by IngenuityArts** [comment]
For the SriMil approach, the dynamic that was of interest was the inequality (Gini coefficient) dynamic evolving over time (whether increasing, decreasing, staying static, chaotic). Our hypothesis that the presence of proto-institutions would increase inequality preceded the choice of what to watch. We weren't looking for an equilibrium but by running a certain number of iterations - say 1000 or whatever - we expected that if proto-institutions supported our hypothesis, we would see that dynamic emerge in some notable way within those iterations. While the number of iterations is not formally determined, it is a reasonable, experimental parameter for initial, exploratory steps.

**Comment by IngenuityArts** [comment]
One of the problems with Gini is the averaging effect, that we may be watching systemic stasis while there is interesting variation at lower scales. This was part of the interest in adding an ability to see individual agents and proto-institutions evolving over time. Systemic averaging is well known for the way it masks these sub-dynamics and is something a proto-anything needs to pay attention to. 

As per Jeff's note, visualization is not only nice, it provides types of information that numeric summaries and statistics may not. We all use plots, histograms, and other tools to take a look at new data sets for exactly the information dense way that images (both static and dynamic) reveal interactions otherwise masked.

**Comment by jeffg828** [comment]
Can an agent enter into more than one proto institution?

**Comment by srikanthmudigonda** [comment]
The time period for which a simulation runs seems to be determined using a combination of heuristics drawn from theory and practical considerations. 

Theoretical considerations include such things as "equilibrium" or oscillations among a set of states; practical considerations include such things as time taken to execute a certain number of time cycles and, where feasible, replications. In the model we presented at CSSSA 2018, we used the latter consideration - we had to set 20 time periods as a reasonably (arguably) sufficient time to see how the evolution happened. 

The execution time for both a single run and replications X number of parameter combinations was the constraint that informed our choice of 20 time periods. We recognize that that choice is possibly pretty constraining in seeing (a) whether an equilibrium or an oscilation-among-states phenomenon occurs, or (b) whether in certain situations the sugarscape goes barren and/or all agents die. 

Once the port to Julia happens, I hope that we can explore these theoretically more interesting phenomena.

**Comment by divilian** [comment]
> Can an agent enter into more than one proto institution?

In [the current SPECnet model](https://cssas.pubpub.org/pub/1y9vib1p/draft), no. Once an agent joins a proto, it never joins another. (Eventually protos will be able to dissolve under certain conditions, after which its agents will be able to join another.)

**Comment by venkatachalapathy** [comment]
@jeffg828 , @WheezePuppet @srikanthmudigonda and @IngenuityArts , I feel like an annoying stack exchange moderator but the current discussion is off-topic. It is better discussed on other issues like #16, #14 and #13 .

**Comment by divilian** [comment]
I have added a wealth histogram animation (which is also visible via caladan). It doesn't color code by proto-vs-non-proto agents yet, which is my next task. This animation is also not linked to the other in a frame-by-frame fashion as one might expect; that would be a cool feature. Maybe @will-nehrboss has another amazing idea for how to make that work.

Leaving this issue open for now, until I implement color coding.

**Comment by venkatachalapathy** [comment]
This is related to end point of simulation question. I created a new issue regarding that #29
